### PR TITLE
Added more escaping characters for HTML

### PR DESCRIPTION
**Pull request by Ryuuzakis**
Hi.

I've parsed a list of html escape characters (that you can find [here](http://www.theukwebdesigncompany.com/articles/entity-escape-characters.php)), and added them to the html.cr file.

Hope it's good :smile:

### CODE REVIEW

**Comment by marceloboeira** [comment]
@Ryuuzakis 

Tests are failing, take a look: https://travis-ci.org/manastech/crystal/jobs/110107440
```
crystal spec ./spec/std/html_spec.cr:6 # HTML .escape does not change a safe string
crystal spec ./spec/std/html_spec.cr:12 # HTML .escape escapes dangerous characters from a string
```

**Comment by asterite** [comment]
I'm not sure we need to escape those characters. The current implementation contains problematic characters that you always want to escape in HTML, like `<`, `>`, `'`, `"` and `&`. I think I needed to escape other chars, like accented spanish chars, when I was programming in PHP maybe 8 years ago, but I think now that's not needed anymore.

**Comment by Ryuuzakis** [comment]
Oh right ! I forgot to launch the tests, my bad.

So you say we should leave it as it was ? Or maybe remove some unneeded escaped characters ? I just lazily let them all, but I could sort them a bit to make them pass the tests.

**Comment by asterite** [comment]
I think the current list is fine. But I honestly don't know when and what characters need to always be escaped in HTML.

**Comment by Ryuuzakis** [comment]
Okay, so I've searched a bit, and it seems that we need a little more characters escaping that the set there already is.

According to [this blog article](http://wonko.com/post/html-escaping), which refers and explains content from the [OWASP](https://www.owasp.org/index.php/XSS_%28Cross_Site_Scripting%29_Prevention_Cheat_Sheet#XSS_Prevention_Rules), you need more characters to prevent XSS injections.

I will update my branch by limiting the characters to what is really needed when I have some time to spare :)

File: src/html.cr [hunk 1]
```diff
@@ -1,11 +1,372 @@
 module HTML
-  SUBSTITUTIONS = {
-    '&'  => "&amp;",
-    '\"' => "&quot;",
-    '\'' => "&apos;",
-    '<'  => "&lt;",
-    '>'  => "&gt;",
-  }
+SUBSTITUTIONS = {
```
Comment (luislavena) [inline src/html.cr]: Looks like indentation miss? `SUBSTITUTIONS` should be nested (by 2 spaces) inside `HTML` and all the elements in the hash also nested (by 2 spaces) from `SUBSTITUTIONS` definition.

You can use Crystal's `format` to correct that (`crystal tool format src/html.cr`)

**Comment by luislavena** [comment]
Perhaps we should look at other implementations of HTML escaping before decide? Basically the substitution list looks too big now :stuck_out_tongue_winking_eye:

**Comment by ysbaddaden** [comment]
@Ryuuzakis goot catch. We should indeed escape some more characters to be (almost) safe.

File: src/html.cr (hunk 1 above)
Comment (Ryuuzakis) [inline src/html.cr]: Thanks for the tip, I'll look into it !

**Comment by Ryuuzakis** [comment]
I've done the modifications I talked about earlier, and added a new test case for the escape method.
Telle me what you think about it :)

**Comment by asterite** [comment]
@Ryuuzakis Thanks!

Do we need to convert spaces to `&nbsp;`, though? I also don't understand why the spaces are not converted in your spec...

**Comment by Ryuuzakis** [comment]
The space that is protected is the non-breaking space, this is why the classical space is not converted.

**Comment by asterite** [comment]
@Ryuuzakis I see. I think it's very confusing if we leave it like that. Another confusing case is the single quote, it looks like `'''`, which looks like a typo. Better change `'''` to `'\''` and the non-breaking space to ``\u{a0}'` or `160.chr`.

**Comment by Ryuuzakis** [comment]
@asterite I agree with you, your solution was better, so I just implemented it. I've also added a test case to test the non-breaking spaces and normal spaces.

File: src/html.cr [hunk 2]
```diff
@@ -1,10 +1,24 @@
...
+    '>'      => "&gt;",
+    '<'      => "&lt;",
+    '+'      => "&#43;",
+    '@'      => "&#64;",
+    '['      => "&#91;",
+    ']'      => "&#93;",
+    '`'      => "&#96;",
+    '{'      => "&#123;",
+    '}'      => "&#125;",
+    '\u{a0}' => "&nbsp",
```
Comment (luislavena) [inline src/html.cr]: Perhaps missing `;` at the end?

Comment (asterite) [inline src/html.cr]: @luislavena Good catch

Comment (Ryuuzakis) [inline src/html.cr]: The `;` is not needed for it to work. Do you want me to add it, for consistency purpose ?

Comment (luislavena) [inline src/html.cr]: @Ryuuzakis afaik all HTML entities must be enclosed between `&` and `;`, at least that is what W3C says, if that works in some browsers that will be fun.

Either way, I think for consistency, yes, let's have them according to W3C.

Thank you.

Comment (Ryuuzakis) [inline src/html.cr]: @luislavena Guess you've got a valid point. It's done :)

**Comment by asterite** [comment]
@Ryuuzakis Thank you! We are almost done here. The indentation is wrong in the spec file. Could you run `crystal tool format` to fix that?

**Comment by Ryuuzakis** [comment]
Done !

**Comment by ysbaddaden** [comment]
Thank you!

**Pull request merged**